### ISSUE TITLE
`Ecto.Changeset.put_change/3` puts change wrapped in changeset for assocs in embedded schemas

### ISSUE THREAD

**Issue by florius0** [elixir-ecto/ecto#4361]
### Elixir version

Elixir 1.15.6

### Database and Version

PostgreSQL 15.4

### Ecto Versions

3.11.1 and 3.10.3

### Database Adapter and Versions (postgrex, myxql, etc)

postgrex 0.17.4

### Current behavior

When using `Ecto.Changeset.put_change/3`/`Ecto.Changeset.put_change/3` of association of embedded schema, `Repo.insert`'s result is returned with the change being changeset instead of the actual change, i.e.:

```elixir
import Ecto.Changeset

defmodule A do
  use Ecto.Schema
    
  schema "a" do
    embeds_one :b, B
  end
  
  def changeset(a, attrs) do
    a
    |> cast(attrs, [])
    |> cast_embed(:b, with: &B.changeset/2)
  end
end

defmodule B do
  use Ecto.Schema
  
  embedded_schema do
    belongs_to :c, C
  end
  
  def changeset(b, attrs) do
    b
    |> struct(c: nil) # Another bug which I'll report later, related to `Relation.load!` behavior with embedded schemas (they lack `__meta__` field)
    |> put_assoc(:c, Repo.get(C, 1))
  end
end

defmodule C do
  use Ecto.Schema

  schema "c" do
    # ...
  end
end

%A{}
|> cast(%{a: %{b: %{c: :smth}}})
|> Repo.insert()
```

will result in

```
{
  :ok,
  %A{
    b: %B{
      c: %Ecto.Changeset{....}
    }
  }
}
```

### Expected behavior

```
{
  :ok,
  %A{
    b: %B{
      c: %C{...}
    }
  }
}
```

**Comment by greg-rychlewski** [comment]
Thanks for the report.

I think we need to recurse through the embeds to perform the insert. I'm pretty sure that nested association is not being inserted at all. Are you able to confirm?

**Comment by florius0** [comment]
> I'm pretty sure that nested association is not being inserted at all

Nested association itself is not being inserted, only the "fk" field is

**Comment by greg-rychlewski** [comment]
I'm trying to reproduce locally but it's a bit difficult. Do you have an example that I can reproduce? `put_assoc` cannot accept a struct here:

```elixir
def changeset(b, attrs) do
    b
    |> struct(c: nil) # Another bug which I'll report later, related to `Relation.load!` behavior with embedded schemas (they lack `__meta__` field)
    |> put_assoc(:c, Repo.get(C, 1))
  end
``` 

Not sure what you did instead. Thanks.

**Comment by florius0** [comment]
I'll provide you with full reproduction shortly

**Comment by josevalim** [comment]
So it seems associations in embeds are pretty much read-only? That seems to be the intuition here too: https://github.com/elixir-ecto/ecto/pull/3967

**Comment by greg-rychlewski** [comment]
That is what I understand as well. 

When I went through the history I found that embeds did not support associations initially, then it was considered due to NoSQL databases here: https://github.com/elixir-ecto/ecto/issues/1837. Then I believe it was dropped for a while until someone submitted a PR for preloading only. 

I *think* it could be extended to writes, with the caveat about the metadata listed above. But would need to try implementing it to be sure I am not missing any edge cases.
